starrSeq/normalize.go

```diff
@@ -0,0 +1,190 @@
+package starrSeq
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/sam"
+	"strings"
+)
+
+// InputNormFactor is a custom struct that scores input-normalization factors for STARR-seq constructs
+type InputNormFactor struct {
+	Construct  string
+	NormFactor float64
+}
+
+// GfpNormalize iterates over a single-cell counts map and applies a GFP normalization factor.
+func GfpNormalize(countsMap map[string]float64, normFactor float64) {
+	var counts float64
+	for i := range countsMap {
+		counts, _ = countsMap[i]
+		countsMap[i] = counts * normFactor
+	}
+}
+
+// CalculateGfpNormFactor takes in a map of cellType-gfpReads and returns a map with cellType-gfpNormalizationValue. Normalization value is calculated by the percentage of GFP reads if all
+// cell types were equal, and dividing it by the observed percentage of GFP reads in a given cell type
+func CalculateGfpNormFactor(clusterGFP map[string]int) map[string]float64 {
+	var totalCounts, clusterCounts int
+	var actualPerc float64
+	gfpNormMap := make(map[string]float64)
+	idealPerc := 1.0 / float64(len(clusterGFP))
+	for i := range clusterGFP {
+		clusterCounts, _ = clusterGFP[i]
+		totalCounts = clusterCounts + totalCounts
+	}
+	for i := range clusterGFP {
+		actualPerc = float64(clusterGFP[i]) / float64(totalCounts)
+		gfpNormMap[i] = idealPerc / actualPerc
+	}
+	return gfpNormMap
+}
+
+// GetNcAvg takes in the input normalized matrix of single cell data, the list of negative control sequences and an empty slice of slices and will return the average negative control reads for each
+// cell type
+func GetNcAvg(matrix []scStarrSeqMatrix, ncSlice []string, ncVals [][]float64) []float64 {
+	var found bool
+	var currCluster string
+	var prevCluster string
+	var ncAvg []float64
+
+	clusterIdx := 0
+	for p, i := range matrix {
+		found = false
+		currCluster = i.cluster
+		if p > 0 && currCluster != prevCluster {
+			clusterIdx++
+		}
+		for _, j := range ncSlice {
```

Comment:
Note to self: for each element of matrix, check if its "construct" is matches any one of the "ncSlice constructs"